core/state/statedb.go

```diff
@@ -866,3 +866,24 @@ func (s *StateDB) Commit(deleteEmptyObjects bool) (common.Hash, error) {
 	}
 	return root, err
 }
+
+// Clean returns true if the statedb does not have any pending modifications
+func (s *StateDB) IsClean() bool {
+	// It is dirty is any single of the conditions holds
+	return !(s.txIndex > 0 ||
+		s.thash != (common.Hash{}) ||
+		s.bhash != (common.Hash{}) ||
+		len(s.stateObjectsDirty) > 0 ||
+		len(s.stateObjectsPending) > 0 ||
+		len(s.journal.dirties) > 0 ||
+		len(s.journal.entries) > 0)
+}
```

Comment:
One (simpler) alternative might be to drill one layer deeper and add a function to `trie.Trie` that will allow for checking if the trie is dirty based on the `t.unhashed` value. https://github.com/celo-org/celo-blockchain/blob/5f046188332511f2974357653a84e3f0008a58fe/trie/trie.go#L54